blockchain/state_manager/src/lib.rs

```diff
@@ -858,77 +852,59 @@ where
             .unwrap_or(Ok(None))
     }
 
-    async fn check_search(
+    fn check_search(
         &self,
-        current: &Tipset,
+        mut current: Arc<Tipset>,
         (message_from_address, message_cid, message_sequence): (&Address, &Cid, &u64),
-    ) -> Result<Option<(Arc<Tipset>, Receipt)>, Result<Arc<Tipset>, Error>> {
-        if current.epoch() == 0 {
-            return Ok(None);
-        }
-        let state = StateTree::new_from_root(self.blockstore(), current.parent_state())
-            .map_err(|e| Err(Error::State(e.to_string())))?;
-
-        if let Some(actor_state) = state
-            .get_actor(message_from_address)
-            .map_err(|e| Err(Error::State(e.to_string())))?
-        {
-            if actor_state.sequence == 0 || actor_state.sequence < *message_sequence {
+    ) -> Result<Option<(Arc<Tipset>, Receipt)>, Error> {
+        loop {
```

Comment:
Why was the `loop` construct added? Looks more complex than just moving from `async` to `sync` here.

Reply:
The `loop` used to be in `search_back_for_message`. I merged the code because the return type of `check_search` was too complicated for me to understand: `Result<Option<(Arc<Tipset>, Receipt)>, Result<Arc<Tipset>, Error>>`

By moving the loop into `check_search`, the return type can be simplified to: `Result<Option<(Arc<Tipset>, Receipt)>, Error>`

The diff is a bit convoluted but the code is more-or-less the same.